spec/services/facility_stats_service_spec.rb

```diff
@@ -0,0 +1,144 @@
+    end
+
+    it "accurately tallies stats for facilities by size and period" do
+      small_facilities = setup_for_size("small", december - 5.months)
+      medium_facilities = setup_for_size("medium", december - 4.months)
+      large_facilities = setup_for_size("large", december - 3.months)
+      refresh_views
+      all_facilities = small_facilities + medium_facilities + large_facilities
+      subject = FacilityStatsService.new(accessible_facilities: all_facilities, retain_facilities: all_facilities,
+                                         ending_period: period, rate_numerator: "controlled_patients")
+      subject.call
+
+      # all numbers except cumulative_registrations appear in data 3 months after they're recorded
+      expect(stats_of_type(subject.stats_by_size["small"], "controlled_patients")).to eq [0, 0, 0, 2, 0, 0]
+      expect(stats_of_type(subject.stats_by_size["small"], "adjusted_registrations")).to eq [0, 0, 0, 3, 3, 3]
+      expect(stats_of_type(subject.stats_by_size["small"], "cumulative_registrations")).to eq [3, 3, 3, 3, 3, 3]
+      expect(stats_of_type(subject.stats_by_size["small"], "controlled_patients_rate")).to eq [0, 0, 0, 67, 0, 0]
+
+      expect(stats_of_type(subject.stats_by_size["medium"], "controlled_patients")).to eq [0, 0, 0, 0, 2, 0]
+      expect(stats_of_type(subject.stats_by_size["medium"], "adjusted_registrations")).to eq [0, 0, 0, 0, 3, 3]
+      expect(stats_of_type(subject.stats_by_size["medium"], "cumulative_registrations")).to eq [0, 3, 3, 3, 3, 3]
+      expect(stats_of_type(subject.stats_by_size["medium"], "controlled_patients_rate")).to eq [0, 0, 0, 0, 67, 0]
+
+      expect(stats_of_type(subject.stats_by_size["large"], "controlled_patients")).to eq [0, 0, 0, 0, 0, 2]
+      expect(stats_of_type(subject.stats_by_size["large"], "adjusted_registrations")).to eq [0, 0, 0, 0, 0, 3]
+      expect(stats_of_type(subject.stats_by_size["large"], "cumulative_registrations")).to eq [0, 0, 3, 3, 3, 3]
+      expect(stats_of_type(subject.stats_by_size["large"], "controlled_patients_rate")).to eq [0, 0, 0, 0, 0, 67]
+    end
+
+    it "sets stats for all accessible facilities but only sets facilities data for retained facilities" do
+      small_facilities = setup_for_size("small")
+      large_facilities = setup_for_size("large")
+      accessible_facilities = small_facilities + large_facilities
+      subject = FacilityStatsService.new(accessible_facilities: accessible_facilities, retain_facilities: large_facilities,
+                                         ending_period: period, rate_numerator: "controlled_patients")
+      subject.call
+      expect(subject.stats_by_size.keys).to match_array(["small", "large"])
+      expect(subject.facilities_data.keys).to match_array(large_facilities.map(&:name))
+    end
+
+    it "processes data for controlled_patients" do
+      subject = FacilityStatsService.new(accessible_facilities: [small_facility1], retain_facilities: [small_facility1],
```

Comment:
I would suggest adding a class `call` method that forwards args to the initializer, then you can just do 

```
result = FacilityStatsService.call(accessible_facilities: [small_facility1], retain_facilities: [small_facility1]....)
```

We follow that type of pattern quite a bit, see https://github.com/simpledotorg/simple-server/blob/8014f0f5da0187a66202a1d3475f642db0357a0a/app/services/reports/region_service.rb#L10-L12 for example.

Reply:
Yeah, that's a common pattern in ruby. Sounds good.